src/basics.ml

```diff
@@ -166,3 +166,22 @@ let term_of_rhs : rule -> term = fun r ->
     TE_Some(Bindlib.unbox (Bindlib.bind_mvar vars p))
   in
   Bindlib.msubst r.rhs (Array.mapi fn r.vars)
+
+(** [init_list n f] builds a list with n elements where each element
+    is given by the function f, in increasing order, ie, it builds
+    [f 0; f 1; ...; f (n-1)] *)
+let init_list : int -> (int -> 'a) -> 'a list = fun n f ->
+  let rec init_list_aux : int -> ('a list) -> 'a list = fun i acc ->
+    if i>=n then acc
+    else init_list_aux (i+1) ((f i)::acc)
+  in List.rev (init_list_aux 0 [])
```

Comment:
Can't we define this function without using List.rev?

Reply:
If we don't want to reverse it at the very end, then we need to build it directly in the right order, using an "append" (i.e, acc @ [f i]) instead of a "cons" (i.e, (f i)::acc) in every recursive call.
Reversing the list will be in O(n) (n being the input given by the user, which is the size of the list produces by the auxiliary function init_list_aux), but appending every time during every recursive call (instead of cons-ing) would lead would give O(1) + O(2) + ... + O(n) which is strictly greater than O(n). So I'd be in favor of letting the reverse afterwards.

Reply:
Why not start with f (n-1) and decrease down to 0 ?
```suggestion
  let rec init_list_aux : int -> ('a list) -> 'a list = fun i acc ->
    if i < 0 then acc
    else init_list_aux (i-1) (f i :: acc)
  in init_list_aux (n-1) []
```

Reply:
Good point, but the answer is because in the case of the function doing side-effects, these side effects would happen in the wrong order. We need the computations of the (f i) (i from 0 to n) to happen precisely from 0 to n.

Reply:
Note that tail recursion is actually harmful here, since this function will only be called with very small integers. So you can avoid the reversal using the following implementation.
```ocaml
let init_list : int -> (int -> 'a) -> 'a list = fun n f ->
  if n < 0 then invalid_arg "Basics.init_list";
  let rec init_list i = if i >= n then [] else f i :: init_list (i+1) in
  init_list 0
```

Reply:
By the way, this function should really be in `Extra.List` under the name `init`.